pilot/pkg/serviceregistry/kube/controller/ambient/multicluster.go

```diff
@@ -336,11 +298,114 @@ func (a *index) buildGlobalCollections(
 		opts,
 	)
 
-	WorkloadAddressIndex := krt.NewIndex[networkAddress, model.WorkloadInfo](GlobalWorkloads, "networkAddress", networkAddressFromWorkload)
-	WorkloadServiceIndex := krt.NewIndex[string, model.WorkloadInfo](GlobalWorkloads, "service", func(o model.WorkloadInfo) []string {
+	workloadNetworkServiceIndex := krt.NewIndex[string, model.WorkloadInfo](GlobalWorkloads, "network;service", func(o model.WorkloadInfo) []string {
+		res := make([]string, 0, len(o.Workload.Services))
+		for svc := range o.Workload.Services {
+			res = append(res, strings.Join([]string{o.Workload.Network, svc}, ";"))
+		}
+		return res
+	})
+
+	coalescedWorkloads := krt.NewManyCollection(
+		workloadNetworkServiceIndex.AsCollection(),
+		func(ctx krt.HandlerContext, i krt.IndexObject[string, model.WorkloadInfo]) []model.WorkloadInfo {
+			parts := strings.Split(i.Key, ";")
+			if len(parts) != 2 {
+				log.Errorf("Invalid key %s for SplitHorizonWorkloads, expected <network>;<service>", i.Key)
+				return nil
+			}
+			networkID := network.ID(parts[0])
+			localNetwork := a.Network(ctx).String()
+			if networkID.String() == localNetwork {
+				// We don't coalesce workloads for the local network
+				return nil
+			}
+
+			svcName := parts[1]
+			svc := krt.FetchOne(ctx, a.services.Collection, krt.FilterKey(svcName))
+			if svc == nil {
+				log.Errorf("Failed to find service %s to coalesce workloads", svcName)
+				return nil
+			}
+			if svc.Scope != model.Global {
+				return nil
+			}
+
+			capacity := uint32(0)
+			for _, wl := range i.Objects {
+				if wl.Workload.GetCapacity().GetValue() == 0 {
+					capacity++
+				} else {
+					capacity += wl.Workload.Capacity.GetValue()
+				}
+			}
+			gws := LookupNetworkGateway(ctx, networkID, a.networks.NetworkGateways, a.networks.GatewaysByNetwork)
+			meshCfg := krt.FetchOne(ctx, a.meshConfig.AsCollection())
+			if meshCfg == nil {
+				log.Errorf("Failed to find mesh config for network %s", networkID)
+				return nil
+			}
+			if len(gws) == 0 {
+				log.Warnf("No network gateway found for network %s", networkID)
+				return nil
+			}
+			if len(gws) > 1 {
+				log.Warnf("Multiple gateways found for network %s, using the first one", networkID)
+			}
+			gw := gws[0]
+			wi := a.createSplitHorizonWorkload(svcName, svc.Service, &gw, capacity, meshCfg)
+			wi = ptr.Of(precomputeWorkload(*wi))
+			return []model.WorkloadInfo{*wi}
+		}, opts.WithName("remoteCoalesedWorkloads")...,
+	)
+	networkLocalWorkloads := krt.NewCollection(GlobalWorkloads, func(ctx krt.HandlerContext, i model.WorkloadInfo) *model.WorkloadInfo {
+		if i.Workload.Network == a.Network(ctx).String() {
```

Comment:
nit(style): exiting early is cleaner, more idiomatic go: https://danp.net/posts/reducing-go-nesting/